Promote admission webhook API to v1

liggitt opened the pull request:
**What type of PR is this?**
/kind api-change
/kind feature

**What this PR does / why we need it**:
* Promotes the {Validating,Mutating}WebhookConfiguration types to v1
* `failurePolicy` default changed from `Ignore` to `Fail` for v1
* `matchPolicy` default changed from `Exact` to `Equivalent` for v1
* `timeout` default changed from `30s` to `10s` for v1
* `sideEffects` default value is removed and the field made required for v1
* `admissionReviewVersions` default value is removed and the field made required for v1
* webhook names required to be unique when creating v1 configuration objects (https://github.com/kubernetes/kubernetes/issues/78510)
* Adds unit tests for new defaults
* Adds unit tests for new validation

Fixes https://github.com/kubernetes/kubernetes/issues/78510


**Special notes for your reviewer**:

* The first commit just copies the existing types, changes package names, and finds/replaces v1beta1 with v1
* The subsequent commits change or remove a specific default in the v1 API, or tighten validation in the v1 API
* The last commit is generated files

**Does this PR introduce a user-facing change?**:
```release-note
The `MutatingWebhookConfiguration` and `ValidatingWebhookConfiguration` APIs have been promoted to `admissionregistration.k8s.io/v1`:
* `failurePolicy` default changed from `Ignore` to `Fail` for v1
* `matchPolicy` default changed from `Exact` to `Equivalent` for v1
* `timeout` default changed from `30s` to `10s` for v1
* `sideEffects` default value is removed, and the field made required, and only `None` and `NoneOnDryRun` are permitted for v1
* `admissionReviewVersions` default value is removed and the field made required for v1 (supported versions for AdmissionReview are `v1` and `v1beta1`)
* The `name` field for specified webhooks must be unique for `MutatingWebhookConfiguration` and `ValidatingWebhookConfiguration` objects created via `admissionregistration.k8s.io/v1`
The `admissionregistration.k8s.io/v1beta1` versions of `MutatingWebhookConfiguration` and `ValidatingWebhookConfiguration` are deprecated and will no longer be served in v1.19.
```

/sig api-machinery
/cc @sttts @jpbetz @roycaihw

liggitt committed:
Copy v1beta1 to v1 admission registration types

liggitt committed:
Install/register v1 admission registration types

liggitt committed:
Change default failurePolicy to Fail in v1

liggitt committed:
Change default matchPolicy to Equivalent in v1

liggitt committed:
Remove default sideEffects in v1, make required in validation

liggitt committed:
Remove default admissionReviewVersions in v1, make required in validation

liggitt committed:
Generated

liggitt commented:
/retest

liggitt commented:
/retest

liggitt commented:
/priority important-soon

liggitt committed:
Add defaulting tests

liggitt committed:
Require webhook names to be unique in v1

liggitt committed:
Add mutatingwebhook validation tests

liggitt commented:
/assign @smarterclayton 
for API review

smarterclayton commented on pkg/apis/admissionregistration/validation/validation_test.go:
```diff
@@ -344,6 +358,17 @@ func TestValidateValidatingWebhookConfiguration(t *testing.T) {
 			expectedError: `webhooks[0].failurePolicy: Unsupported value: "other": supported values: "Fail", "Ignore"`,
 		},
+		{
+			name: "SideEffects are required",
```
Gah, grammar.  It's fine though.

sttts commented on pkg/apis/admissionregistration/validation/validation.go:
```diff
@@ -275,6 +278,9 @@ func validateMutatingWebhook(hook *admissionregistration.MutatingWebhook, fldPat
 		allErrors = append(allErrors, field.NotSupported(fldPath.Child("matchPolicy"), *hook.MatchPolicy, supportedMatchPolicies.List()))
 	}
+	if hook.SideEffects == nil {
+		allErrors = append(allErrors, field.Required(fldPath.Child("sideEffects"), fmt.Sprintf("must specify one of %v", supportedSideEffectClasses.List())))
```
unify with create validation error

liggitt replied:
not sure what you mean by this

sttts replied:
you have `%v` once with a slice and `strings.Join` everywhere else for the possible values.

liggitt replied:
oh, fixed in later commit, will collapse

sttts commented on pkg/registry/admissionregistration/validatingwebhookconfiguration/strategy.go:
```diff
@@ -63,7 +65,12 @@ func (validatingWebhookConfigurationStrategy) PrepareForUpdate(ctx context.Conte
-	return validation.ValidateValidatingWebhookConfiguration(obj.(*admissionregistration.ValidatingWebhookConfiguration))
+	var groupVersion schema.GroupVersion
+	if requestInfo, found := genericapirequest.RequestInfoFrom(ctx); found {
+		groupVersion = schema.GroupVersion{Group: requestInfo.APIGroup, Version: requestInfo.APIVersion}
```
if we panic in the else case, what breaks? Only tests?

liggitt replied:
I wanted to make that any internal code paths calling this would get the tight validation. Only the invocation that explicitly comes in via v1beta1 gets the looser validation. This is the same thing we do in other places:

https://github.com/kubernetes/kubernetes/blob/79aab0b90d9564dd3ae28b1764608c6dd0aa09f7/pkg/registry/storage/volumeattachment/strategy.go#L78-L90

sttts commented on pkg/apis/admissionregistration/validation/validation.go:
```diff
@@ -194,28 +195,50 @@ func validateAdmissionReviewVersions(versions []string, requireRecognizedVersion
-func ValidateValidatingWebhookConfiguration(e *admissionregistration.ValidatingWebhookConfiguration) field.ErrorList {
-	return validateValidatingWebhookConfiguration(e, true)
+// ValidateValidatingWebhookConfiguration validates a webhook before creation.
+func ValidateValidatingWebhookConfiguration(e *admissionregistration.ValidatingWebhookConfiguration, gv schema.GroupVersion) field.ErrorList {
```
s/gv/requestGV/

liggitt replied:
done

sttts commented on pkg/apis/admissionregistration/validation/validation.go:
```diff
@@ -194,28 +195,50 @@ func validateAdmissionReviewVersions(versions []string, requireRecognizedVersion
 		allErrors = append(allErrors, validateAdmissionReviewVersions(hook.AdmissionReviewVersions, requireRecognizedVersion, field.NewPath("webhooks").Index(i).Child("admissionReviewVersions"))...)
+		if requireUniqueWebhookNames && len(hook.Name) > 0 {
+			if hookNames.Has(hook.Name) {
+				allErrors = append(allErrors, field.Duplicate(field.NewPath("webhooks").Index(i).Child("name"), hook.Name))
```
what's the reason we don't have a field path arguments in these validation funcs?

sttts replied:
ok, it's the root object. nvmd. Would be nil anyway.

sttts commented on pkg/apis/admissionregistration/validation/validation.go:
```diff
@@ -403,10 +426,52 @@ func validatingHasAcceptedAdmissionReviewVersions(webhooks []admissionregistrati
+	var requireUniqueWebhookNames bool
+	if gv == (schema.GroupVersion{Group: admissionregistration.GroupName, Version: "v1beta1"}) {
+		// when updating via admissionregistration.k8s.io/v1beta1, do not require unique names for compatibility
+		requireUniqueWebhookNames = false
```
it's false anyway as zero value.

liggitt replied:
simplified

liggitt commented:
addressed comments in https://github.com/kubernetes/kubernetes/compare/4c57e6ee0555e15cc67ba0069b4015672dec10d7..8500cb278f873b75923dd5a2e63ec9cd1d63d4e8

roycaihw commented on pkg/apis/admissionregistration/v1/defaults.go:
```diff
@@ -0,0 +1,92 @@
+	}
+	if obj.TimeoutSeconds == nil {
+		obj.TimeoutSeconds = new(int32)
+		*obj.TimeoutSeconds = 30
```
let's change the default timeout to [10 seconds](https://github.com/kubernetes/enhancements/blob/master/keps/sig-api-machinery/00xx-admission-webhooks-to-ga.md#timeout-configuration) in v1

liggitt replied:
should we go even shorter? a 10 second hiccup in API processing is still really long

roycaihw replied:
the KEP gave 10s as an example, so I think it doesn't have to be 10s. I remember in previous meeting we even talked about something like 1s.

cc @lavalamp WDYT?

liggitt replied:
I think 10s is probably a good balance. Updated.

roycaihw commented on staging/src/k8s.io/api/admissionregistration/v1/types.go:
```diff
@@ -0,0 +1,551 @@
+	// and "rules" only included `apiGroups:["apps"], apiVersions:["v1"], resources: ["deployments"]`,
+	// a request to apps/v1beta1 or extensions/v1beta1 would be converted to apps/v1 and sent to the webhook.
+	//
+	// Defaults to "Exact"
```
doc: Defaults to "Equivalent"

liggitt replied:
fixed

roycaihw commented on staging/src/k8s.io/api/admissionregistration/v1/generated.proto:
```diff
@@ -0,0 +1,479 @@
+  // rejected by a future step in the admission change and the side effects therefore need to be undone.
+  // Requests with the dryRun attribute will be auto-rejected if they match a webhook with
+  // sideEffects == Unknown or Some.
+  optional string sideEffects = 6;
```
is something wrong with the generator? I thought sideEffects should be required

roycaihw replied:
nevermind, this seems to be intended

liggitt committed:
Change default timeout to 10 seconds

liggitt commented:
/retest

sttts commented:
lgtm
/approve

/assign @smarterclayton 

for final lgtm

smarterclayton commented:
/lgtm

roycaihw commented:
/area admission-control
